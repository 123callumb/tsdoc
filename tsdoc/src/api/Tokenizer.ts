import { TextRange } from './TextRange';
import { Character } from '../internal/Character';

/**
 * Distinguishes different types of Token objects.
 */
export enum TokenKind {
  /**
   * A token representing the end of the input.
   */
  EndOfInput,
  /**
   * A token representing a sequence of plain text with no special meaning.
   */
  PlainText,
  /**
   * A token representing a virtual newline.  The Token.range will be an empty range,
   * because the actual newline character may be noncontiguous or nonexistent.
   */
  Newline,

  /**
   * A single character that can be used as general punctuation.
   * @remarks
   * The Token.range will always be a string of length 1.
   * When a backslash is interpreted as a literal backslash, it will be returned as
   * PunctuationCharacter instead of BackslashEscapedCharacter.
   */
  PunctuationCharacter,

  /**
   * A backslash followed by another punctuation character.  This disables any special meaning
   * the other character may have (if any).  The backslash itself should not be rendered.
   * @remarks
   * The Token.range will always be a string of length 2.
   */
  BackslashEscapedCharacter
}

/**
 * Represents a contiguous range of characters extracted from one of the doc comment lines
 * being processed by the Tokenizer.  There is a token representing a newline, but otherwise
 * a single token cannot span multiple lines.
 */
export class Token {
  public readonly kind: TokenKind;
  public readonly range: TextRange;
  public readonly line: TextRange;

  public constructor(kind: TokenKind, range: TextRange, line: TextRange) {
    this.kind = kind;
    this.range = range;
    this.line = line;
    if (range.pos < line.pos || range.end > line.end) {
      // Sanity check
      throw new Error('Wrong line');
    }
  }

  public toString(): string {
    return this.range.toString();
  }
}

/**
 * The ICharacter structure uses an empty string to represent the end of the input.
 */
const EOI_CHARACTER: string = '';

/**
 * This helps the Tokenizer keep track of indexes needed to accurately calculated Token.range,
 * given that the buffer index can skip characters when advancing to the next line inside
 * a documentation comment.
 */
interface ICharacter {
  /**
   * Either EOI_CHARACTER (an empty string) to indicate that the end of the input has been reached,
   * or else a string of length one representing a single character extracted from the input buffer.
   */
  value: string;

  /**
   * The buffer index of the character.  This is an index into Tokenizer._buffer.
   */
  index: number;

  /**
   * The input line that this character came from.
   */
  line: TextRange;
}

export class Tokenizer {
  /**
   * The array of doc comment lines being tokenized.
   * These ranges do not include doc comment delimiters such as "/*" or "*".
   */
  public readonly lines: TextRange[];

  // Index into the lines array, indicating the current line being processed.
  private _linesIndex: number;

  // Index into Tokenizer._cachedBuffer, which is the same as this.lines[i].buffer.
  private _bufferIndex: number;

  // If we've reached the end of the input, this is the final Token object.
  private _endToken: Token | undefined;

  // Storage for Token.peekCharacter()
  private _peekedCharacter: ICharacter | undefined;

  // A temporary state used when generating a TokenKind.Newline token
  private _injectedNewline: boolean;

  // To improve performance, this is a cached storage of this.lines[this._linesIndex].
  private _cachedCurrentLine: TextRange | undefined;

  // To improve performance, this is a cached storage of this.lines[i].buffer for all values of i;
  // the lines are assumed to all come from a common buffer.
  private readonly _cachedBuffer: string;

  public constructor(lines: TextRange[]) {
    this.lines = lines;
    this._linesIndex = 0;

    if (this.lines.length === 0) {
      this._cachedCurrentLine = undefined;
      this._cachedBuffer = '';

      this._bufferIndex = 0;
      this._endToken = new Token(TokenKind.EndOfInput, TextRange.empty, TextRange.empty);
    } else {
      this._cachedCurrentLine = this.lines[0];
      this._cachedBuffer = this.lines[0].buffer;

      this._bufferIndex = this._cachedCurrentLine.pos;
      this._endToken = undefined;
    }

    this._peekedCharacter = undefined;
    this._injectedNewline = false;
  }

  /**
   * Extracts and returns the next token from the input stream.
   * @remarks
   * When the end of the input is reached, getToken() will (repeatedly) return
   * the TokenKind.EndOfInput token.
   */
  public getToken(): Token {
    const startCharacter: ICharacter = this._getCharacter();

    switch (startCharacter.value) {
      case EOI_CHARACTER:
        return this._endToken!;

      case '\n':
        return new Token(TokenKind.Newline,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, startCharacter.index),
          startCharacter.line);

      case '\\':
        const next: ICharacter = this._peekCharacter();
        // CommonMark Spec: "Any ASCII punctuation character may be backslash-escaped.'
        // Backslashes before other characters are treated as literal backslashes."
        if (Character.isPunctuation(next.value)) {
          this._getCharacter(); // extract next

          return new Token(TokenKind.BackslashEscapedCharacter,
            TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, next.index + 1),
            startCharacter.line);
        }
        // Treat it as a literal backslash
        return new Token(TokenKind.PunctuationCharacter,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, startCharacter.index + 1),
          startCharacter.line);

      default:
        const lastCharacter: ICharacter = this._skipUntil(['\n']) || startCharacter;
        return new Token(TokenKind.PlainText,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, lastCharacter.index + 1),
          startCharacter.line);
    }
  }

  /**
   * Advances the stream pointer until one of the specified ending characters is reached,
   * or until the end of the input is reached.
   * @returns the last skipped character, or undefined if no characters were skipped
   */
  private _skipUntil(endingCharacters: string[]): ICharacter | undefined {
    let lastSkipped: ICharacter | undefined = undefined;
    while (true) {
      const character: ICharacter = this._peekCharacter();

      if (character.value === EOI_CHARACTER) {
        return lastSkipped;
      }
      if (endingCharacters.indexOf(character.value) >= 0) {
        return lastSkipped;
      }
      lastSkipped = this._getCharacter();
    }
  }

  /**
   * Extracts and returns the next character from the input lines.
   * After each input line is processed, a virtual newline character is returned.
   * When the end of the input is reached, an empty string is returned.
   * The length of the returned string will always be 0 or 1.
   */
  private _getCharacter(): ICharacter {
    if (this._peekedCharacter !== undefined) {
      const character: ICharacter = this._peekedCharacter;
      this._peekedCharacter = undefined;
      return character;
    }

    if (this._endToken) {
      // already reached the end of the input
      return {
        value: EOI_CHARACTER,
        index: this._endToken.range.pos,
        line: this._endToken.line
      };
    }
    if (!this._cachedCurrentLine) {
      // Sanity check
      throw new Error('Tokenizer._currentLine should not be undeifned');
    }

    while (true) {
      if (this._bufferIndex < this._cachedCurrentLine.end) {
        return {
          value: this._cachedBuffer[this._bufferIndex],
          index: this._bufferIndex++,
          line: this._cachedCurrentLine
        };
      }

      // When we reach the logical end of line, we inject a "\n" character (since the
      // real EOL may be disembodied by the doc comment delimiters).
      // Since we don't want to move _bufferIndex, we need a little bit of extra state.
      if (!this._injectedNewline) {
        this._injectedNewline = true;
        return {
          value: '\n',
          index: this._bufferIndex,
          line: this._cachedCurrentLine
        };
      }
      this._injectedNewline = false;

      // Advance to the next line
      ++this._linesIndex;
      if (this._linesIndex >= this.lines.length) {
        // We advanced past the final line
        this._endToken = new Token(TokenKind.EndOfInput,
          this._cachedCurrentLine.getNewRange(this._cachedCurrentLine.end, this._cachedCurrentLine.end),
          this._cachedCurrentLine
        );
        return {
          value: EOI_CHARACTER,
          index: this._endToken.range.pos,
          line: this._endToken.line
        };
      }

      this._cachedCurrentLine = this.lines[this._linesIndex];
      this._bufferIndex = this._cachedCurrentLine.pos;
    }
  }

  /**
   * Similar to _getCharacter(), except it does not advance the (conceptual) input stream pointer.
   */
  private _peekCharacter(): ICharacter {
    if (this._peekedCharacter === undefined) {
      this._peekedCharacter = this._getCharacter();
    }
    return this._peekedCharacter;
  }
}
