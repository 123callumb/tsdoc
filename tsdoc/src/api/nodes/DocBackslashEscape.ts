import { DocNode, DocNodeKind, IDocNodeParameters } from './DocNode';

/**
 * Constructor parameters for {@link DocBackslashEscape}.
 */
export interface IDocBackslashEscapeParameters extends IDocNodeParameters {
}

/**
 * Represents a text character that is preceded by a backslash in order to remove
 * any special meaning that it might have as a TSDoc symbol.
 */
export class DocBackslashEscape extends DocNode {
  /** {@inheritdoc} */
  public readonly kind: DocNodeKind = DocNodeKind.BackslashEscape;

  /**
   * Don't call this directly.  Instead use {@link TSDocParser}
   * @internal
   */
  public constructor(parameters: IDocBackslashEscapeParameters) {
    super(parameters);
  }
}
