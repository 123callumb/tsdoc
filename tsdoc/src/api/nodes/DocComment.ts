import { TextRange } from '../TextRange';
import { ParseError } from '../ParseError';
import { DocNode, DocNodeKind, IDocNodeParameters } from './DocNode';
import { ParserContext } from '../ParserContext';

/**
 * Constructor parameters for {@link DocComment}.
 */
export interface IDocCommentParameters extends IDocNodeParameters {
  parserContext: ParserContext;
}

/**
 * Represents an entire parsed documentation comment.  This is typically the
 * root of the expression tree returned by the parser.
 */
export class DocComment extends DocNode {

  /** {@inheritdoc} */
  public readonly kind: DocNodeKind = DocNodeKind.Comment;

  /**
   * The `sourceRange` indicates the start and end of the original input that was parsed.
   */
  public readonly sourceRange: TextRange;

  /**
   * The text range starting from the opening `/**` and ending with
   * the closing `*\/` delimiter.
   */
  public readonly commentRange: TextRange;

  /**
   * The text ranges corresponding to the lines of content inside the comment.
   */
  public readonly lines: TextRange[];

  /**
   * The nodes that were parsed from the input.
   */
  public readonly nodes: DocNode[];

  /**
   * If any errors occurred during parsing, they are returned in this list.
   */
  public readonly parseErrors: ParseError[] = [];

  /**
   * Don't call this directly.  Instead use {@link TSDocParser}
   * @internal
   */
  public constructor(parameters: IDocCommentParameters) {
    super(parameters);

    this.sourceRange = parameters.parserContext.sourceRange;
    this.commentRange = parameters.parserContext.commentRange;
    this.lines = parameters.parserContext.lines;
    this.nodes = parameters.parserContext.nodes;
    this.parseErrors = parameters.parserContext.parseErrors;
  }
}
