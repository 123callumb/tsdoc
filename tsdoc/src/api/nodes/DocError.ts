import { DocNode, DocNodeKind, IDocNodeParameters } from './DocNode';

/**
 * Constructor parameters for {@link DocError}.
 */
export interface IDocErrorParameters extends IDocNodeParameters {
}

/**
 * Represents a span of comment text that contained invalid markup.
 * The characters should be rendered as plain text.
 */
export class DocError extends DocNode {
  /** {@inheritdoc} */
  public readonly kind: DocNodeKind = DocNodeKind.Error;

  /**
   * Don't call this directly.  Instead use {@link TSDocParser}
   * @internal
   */
  public constructor(parameters: IDocErrorParameters) {
    super(parameters);
  }
}
