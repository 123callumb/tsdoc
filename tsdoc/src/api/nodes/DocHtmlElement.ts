import { DocNode, DocNodeKind, IDocNodeParameters } from './DocNode';

/**
 * Constructor parameters for {@link DocHtmlElement}.
 */
export interface IDocHtmlElementParameters extends IDocNodeParameters {
}

/**
 * Represents a span of comment text that contains an HTML tag
 * conforming to CommonMark syntax rules.
 */
export class DocHtmlElement extends DocNode {
  /** {@inheritdoc} */
  public readonly kind: DocNodeKind = DocNodeKind.HtmlTag;

  /**
   * Don't call this directly.  Instead use {@link TSDocParser}
   * @internal
   */
  public constructor(parameters: IDocHtmlElementParameters) {
    super(parameters);
  }
}
