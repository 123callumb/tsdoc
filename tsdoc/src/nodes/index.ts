
export * from './DocBackslashEscape';
export * from './DocComment';
export * from './DocError';
export * from './DocHtmlElement';
export * from './DocNode';
export * from './DocPlainText';
