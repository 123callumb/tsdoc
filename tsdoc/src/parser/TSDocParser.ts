import { DocComment } from '../nodes';
import { TextRange } from './TextRange';
import { ParserContext } from './ParserContext';
import { LineExtractor } from './LineExtractor';
import { Tokenizer } from './Tokenizer';
import { NodeParser } from './NodeParser';

/**
 * The main API for parsing TSDoc comments.
 */
export class TSDocParser {
  public parseString(text: string): DocComment {
    return this.parseRange(TextRange.fromString(text));
  }

  public parseRange(range: TextRange): DocComment {
    const parserContext: ParserContext = new ParserContext(range);

    LineExtractor.extract(parserContext);

    /**
     * If we can't extract any lines, then skip the other stages
     * of analysis.
     */
    if (parserContext.parseErrors.length === 0) {
      parserContext.tokens = Tokenizer.readTokens(parserContext.lines);
      const nodeParser: NodeParser = new NodeParser(parserContext);
      parserContext.nodes = nodeParser.parse();
    }

    return new DocComment({
      parserContext: parserContext,
      childNodes: parserContext.nodes
    });
  }
}
