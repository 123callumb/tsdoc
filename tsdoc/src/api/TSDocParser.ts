import { Character } from '../internal/Character';

import { DocComment, IDocCommentParameters } from './nodes/DocComment';
import { TextRange } from './TextRange';
import { ParseError } from './ParseError';

// Internal parser state
enum State {
  // Initial state, looking for "/*"
  BeginComment1,
  // Looking for "*" or "* " after "/*"
  BeginComment2,
  // Like State.CollectingLine except immediately after the "/**"
  CollectingFirstLine,
  // Collecting characters until we reach a newline
  CollectingLine,
  // After a newline, looking for the "*" that begins a new line, or the "*/" to end the comment
  AdvancingLine,
  // Exiting the parser loop
  Done
}

/**
 * The main API for parsing TSDoc comments.
 */
export class TSDocParser {
  private static _addError(parameters: IDocCommentParameters, range: TextRange,
    message: string, pos: number, end?: number): void {
    if (!end) {
      if (pos + 1 <= range.buffer.length) {
        end = pos + 1;
      } else {
        end = pos;
      }
    }
    parameters.parseErrors.push(
      new ParseError(message, range.getNewRange(pos, end))
    );
  }

  public parseString(text: string): DocComment {
    return this.parseRange(TextRange.fromString(text));
  }

  public parseRange(range: TextRange): DocComment {
    const parameters: IDocCommentParameters = {
      range: TextRange.empty,
      sourceRange: range,
      lines: [],
      parseErrors: []
    };
    this._parseLines(parameters);

    return new DocComment(parameters);
  }

  /**
   * This step parses an entire code comment from slash-star-star until star-slash,
   * and extracts the content lines.  The lines are stored in IDocCommentParameters.lines
   * and the overall text range is assigned to IDocCommentParameters.range.
   */
  private _parseLines(parameters: IDocCommentParameters): void {
    const range: TextRange = parameters.sourceRange;
    const buffer: string = range.buffer;

    let commentRangeStart: number = 0;
    let commentRangeEnd: number = 0;

    // These must be set before entering CollectingFirstLine, CollectingLine, or AdvancingLine
    let collectingLineStart: number = 0;
    let collectingLineEnd: number = 0;

    let nextIndex: number = range.pos;
    let state: State = State.BeginComment1;

    while (state !== State.Done) {
      if (nextIndex >= range.end) {
        // reached the end of the input
        switch (state) {
          case State.BeginComment1:
          case State.BeginComment2:
            TSDocParser._addError(parameters, range, 'Expecting a "/**" comment', range.pos);
              return;
          default:
            TSDocParser._addError(parameters, range, 'Unexpected end of input', range.pos);
            return;
        }
      }

      const current: string = buffer[nextIndex];
      const currentIndex: number = nextIndex;
      ++nextIndex;
      const next: string = nextIndex < range.end ? buffer[nextIndex] : '';

      switch (state) {
        case State.BeginComment1:
          if (current === '/' && next === '*') {
            commentRangeStart = currentIndex;
            ++nextIndex; // skip the star
            state = State.BeginComment2;
          } else if (!Character.isWhitespace(current)) {
            TSDocParser._addError(parameters, range, 'Expecting a leading "/**"', nextIndex);
            return;
          }
          break;
        case State.BeginComment2:
          if (current === '*') {
            if (next === ' ') {
              ++nextIndex; // Discard the space after the star
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.CollectingFirstLine;
          } else {
            TSDocParser._addError(parameters, range, 'Expecting a leading "/**"', nextIndex);
            return;
          }
          break;
        case State.CollectingFirstLine:
        case State.CollectingLine:
          if (current === '\n') {
            // Ignore an empty line if it is immediately after the "/**"
            if (state !== State.CollectingFirstLine || collectingLineEnd > collectingLineStart) {
              // Record the line that we collected
              parameters.lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.AdvancingLine;
          } else if (current === '*' && next === '/') {
            if (collectingLineEnd > collectingLineStart) {
              parameters.lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = 0;
            collectingLineEnd = 0;
            ++nextIndex; // skip the slash
            commentRangeEnd = nextIndex;
            state = State.Done;
          } else if (!Character.isWhitespace(current)) {
            collectingLineEnd = nextIndex;
          }
          break;
        case State.AdvancingLine:
          if (current === '*') {
            if (next === '/') {
              collectingLineStart = 0;
              collectingLineEnd = 0;

              ++nextIndex; // skip the slash
              commentRangeEnd = nextIndex;
              state = State.Done;
            } else {
              // Discard the "*" at the start of a line

              if (next === ' ') {
                ++nextIndex; // Discard the space after the star
              }

              collectingLineStart = nextIndex;
              collectingLineEnd = nextIndex;
              state = State.CollectingLine;
            }
          } else if (current === '\n') {
            // Blank line
            parameters.lines.push(range.getNewRange(currentIndex, currentIndex));
            collectingLineStart = nextIndex;
          } else if (!Character.isWhitespace(current)) {
            // If the star is missing, then start the line here
            // Example: "/**\nL1*/"

            // (collectingLineStart was the start of this line)
            collectingLineEnd = currentIndex;
            state = State.CollectingLine;
          }
          break;
      }
    }

    parameters.range = range.getNewRange(commentRangeStart, commentRangeEnd);
  }
}
