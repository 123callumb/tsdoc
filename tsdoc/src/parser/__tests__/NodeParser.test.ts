import { TSDocParser } from '../TSDocParser';
import { DocComment, DocNodeKind } from '../../nodes';
import { TokenKind } from '../Tokenizer';

function escape(s: string): string {
  return s.replace(/\n/g, '[n]')
    .replace(/\\/g, '[b]');
}

interface ISnapshotItem {
  nodeKind: string;
  tokens: { tokenKind: string, value: string }[];
}

function matchSnapshot(buffer: string): void {
  const tsdocParser: TSDocParser = new TSDocParser();
  const docComment: DocComment = tsdocParser.parseString(buffer);

  const items: ISnapshotItem[] = [];
  for (const node of docComment.nodes) {
    const item: ISnapshotItem = {
      nodeKind: DocNodeKind[node.kind],
      tokens: node.tokens.map((token) => {
        return {
          tokenKind: TokenKind[token.kind],
          value: escape(token.toString())
        };
      })
    };
    items.push(item);
  }

  expect({
    buffer: escape(buffer),
    lines: docComment.lines.map(x => escape(x.toString())),
    nodes: items
  }).toMatchSnapshot();
}

test('00 Tokenizer simple case', () => {
  matchSnapshot([
    '/**',
    ' * line 1 ', // extra space at end of line
    ' * line 2',
    ' */'
  ].join('\n'));
});

test('01 Tokenizer degenerate cases', () => {
  matchSnapshot('/***/');

  matchSnapshot([
    '/**',
    ' *',
    ' */'
  ].join('\n'));

  matchSnapshot([
    '/**',
    ' ',
    ' ',
    ' */'
  ].join('\n'));

});

test('02 Backslash escapes: positive examples', () => {
  matchSnapshot([
    '/**',
    ' * \\$\\@param',
    ' */'
  ].join('\n'));
});

test('03 Backslash escapes: negative examples', () => {
  matchSnapshot([
    '/**',
    ' * letter: \\A space: \\  end of line: \\',
    ' */'
  ].join('\n'));
});

test('04 Block tags: positive examples', () => {
  matchSnapshot([
    '/**',
    ' * @one @TWO @thRee',
    ' */'
  ].join('\n'));
});

test('05 Block tags: negative examples', () => {
  matchSnapshot([
    '/**',
    ' * @ @@',
    ' * @one@two',
    ' * \\@three',
    ' * @four, @five\\a',
    ' */'
  ].join('\n'));
});
