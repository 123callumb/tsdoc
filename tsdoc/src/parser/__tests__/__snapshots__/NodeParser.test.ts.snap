// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`00 Tokenizer simple case 1`] = `
Object {
  "buffer": "/**[n] * line 1 [n] * line 2[n] */",
  "lines": Array [
    "line 1",
    "line 2",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "line",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "1",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "line",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "2",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`01 Tokenizer degenerate cases 1`] = `
Object {
  "buffer": "/***/",
  "lines": Array [],
  "nodes": Array [],
}
`;

exports[`01 Tokenizer degenerate cases 2`] = `
Object {
  "buffer": "/**[n] *[n] */",
  "lines": Array [
    "",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`01 Tokenizer degenerate cases 3`] = `
Object {
  "buffer": "/**[n] [n] [n] */",
  "lines": Array [
    "",
    "",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`02 Backslash escapes: positive examples 1`] = `
Object {
  "buffer": "/**[n] * [b]$[b]@param[n] */",
  "lines": Array [
    "[b]$[b]@param",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "BackslashEscape",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "$",
        },
      ],
    },
    Object {
      "nodeKind": "BackslashEscape",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "param",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`03 Backslash escapes: negative examples 1`] = `
Object {
  "buffer": "/**[n] * letter: [b]A space: [b]  end of line: [b][n] */",
  "lines": Array [
    "letter: [b]A space: [b]  end of line: [b]",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "letter",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": ":",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
      ],
    },
    Object {
      "nodeKind": "Error",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "A",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "space",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": ":",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
      ],
    },
    Object {
      "nodeKind": "Error",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "Spacing",
          "value": "  ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "end",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "of",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "line",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": ":",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
      ],
    },
    Object {
      "nodeKind": "Error",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`04 Block tags: positive examples 1`] = `
Object {
  "buffer": "/**[n] * @one @TWO @thRee[n] */",
  "lines": Array [
    "@one @TWO @thRee",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "one",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "TWO",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "thRee",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;

exports[`05 Block tags: negative examples 1`] = `
Object {
  "buffer": "/**[n] * @ @@[n] * @one@two[n] * [b]@three[n] * @four, @five[b]a[n] */",
  "lines": Array [
    "@ @@",
    "@one@two",
    "[b]@three",
    "@four, @five[b]a",
  ],
  "nodes": Array [
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "one",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "two",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
    Object {
      "nodeKind": "BackslashEscape",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "three",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "four",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": ",",
        },
        Object {
          "tokenKind": "Spacing",
          "value": " ",
        },
        Object {
          "tokenKind": "OtherPunctuation",
          "value": "@",
        },
        Object {
          "tokenKind": "AsciiWord",
          "value": "five",
        },
      ],
    },
    Object {
      "nodeKind": "Error",
      "tokens": Array [
        Object {
          "tokenKind": "Backslash",
          "value": "[b]",
        },
      ],
    },
    Object {
      "nodeKind": "PlainText",
      "tokens": Array [
        Object {
          "tokenKind": "AsciiWord",
          "value": "a",
        },
        Object {
          "tokenKind": "Newline",
          "value": "",
        },
      ],
    },
  ],
}
`;
